Recyclable cache keys

dhh opened the pull request:
Key-based cache expiration is an incredibly powerful, simple way to do away with the error-ridden ways of manual cache expiration, but it can also be highly wasteful and generate lots of cache trash. 

This happens when you have keys which churn at high velocity, leaving the abandoned keys to be garbage collected with no hint to the fact that they will never be used again. If you cache write volume is high, and you turn over your entire cache allowance frequently enough, this results in cache trash crowding out less-frequently-accessed-but-still-valid keys. Which in turn leads to high cache miss rates.

We can solve this problem by making the keys stable by separating the explicit version. So you can keep a stable key, like "products/1" and an associated version, like "20170202145500", instead of the combined "projects/1-20170202145500" key we've been using so far. This means that no matter how frequently Product/1 is touched, it'll still only write to the same cache key. That's the recycling part here.

This approach is similar to how HTTP caching works. There's a cache key in the form of the URL and then there's a version component in form of the ETAG.

Commit:
Support versioned cache entries

This will form the foundation of recyclable cache keys.

bogdan commented:
I don't like the resulting API I would suppose to be using here:

```
Rails.cache.fetch(["post_preview", post], version: post.updated_at)
# or slightly shorter version:
Rails.cache.fetch(["post_preview", post], version: post)
```

The need to explicitly pass version is sad.

Rails should still need to maintain a nicer API:

``` ruby
Rails.cache.fetch(["post_preview", post])
# with multiple objects too:
Rails.cache.fetch(["post_preview", post, post.author]) 
```

Supposing that any object passed as cache key to have a `cache_version` method besides `cache_key`:

``` ruby
class Post
  def cache_key
     "post/#{id}"
  end
  def cache_version
    updated_at
  end
end
```

dhh commented:
This is a low-level API that is rarely if ever intended to be used
directly. The sugar will be supplied by CacheHelper, which will do the work
of dissecting an active record object to separate stable cache key from
version information.

On Mon, May 15, 2017 at 4:31 PM, Bogdan Gusiev <notifications@github.com>
wrote:

> I don't like the resulting API I would suppose to be using here:
>
> Rails.cache.fetch(["post_preview", post], version: post.updated_at)
> # or slightly shorter version:
> Rails.cache.fetch(["post_preview", post], version: post)
>
> The need to explicitly pass version is sad.
>
> Rails should still need to maintain a nicer API:
>
> Rails.cache.fetch(["post_preview", post])# with multiple objects too:Rails.cache.fetch(["post_preview", post, post.author])
>
> Supposing that any object passed as cache key to have a cache_version
> method besides cache_key:
>
> class Post
>   def cache_key
>      "post/#{id}"
>   end
>   def cache_version
>     updated_at
>   endend
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/rails/rails/pull/29092#issuecomment-301492904>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAAKtespLk5YFxYecejaXGm0KZG_7MQhks5r6GHagaJpZM4NbLH7>
> .
>

bogdan commented:
This API is documented and well maintained. There is no reason for anyone to avoid using this API directly. 
What is an intended way to cache some heavy calculations in model if not through this API?

I use this API directly a lot because we have a project with heavy calculations that are outside of HTTP layer. Also this functionality has nothing to do with HTTP stack so there is no benefit from it being only available in the ActionView level.

dhh commented:
You're free to do that, and if you do, you can use it as it's been
documented. But this caching API should work for sorts of different
purposes, so I prefer to keep it low-level and not overly tied to any
larger of a set of APIs on the other side.

Anyway, once this is completed, you're always free to propose a different
kind of API. But it's not a direction I'm currently interested in pursuing
for this.

On Mon, May 15, 2017 at 4:53 PM, Bogdan Gusiev <notifications@github.com>
wrote:

> This API is documented and well maintained. There is no reason for anyone
> to avoid using this API directly.
> What is an intended way to cache some heavy calculations in model if not
> through this API?
>
> I use this API directly a lot because we have a project with heavy
> calculations that are outside of HTTP layer. Also this functionality has
> nothing to do with HTTP stack so there is no benefit from it being only
> available in the ActionView level.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/rails/rails/pull/29092#issuecomment-301499625>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAAKtfnlMKi5UgIYEkNUo0C5jEQtDEYnks5r6GbugaJpZM4NbLH7>
> .
>

bogdan commented:
Agree this is not 100% backward compatible. But if you would design the system this way, if you would start it from scratch, this is a good sign to review this as a direction to go. We can search for solutions if backward compatibility is the only concern.

dhh commented:
That's just one of the considerations. Anyway, feel free to fork this PR
and work on an alternative implementation. I've noted my skepticism to
having a lot of convention at this level of the stack, but always happy to
consider some real code, even if it's a long shot.

On Mon, May 15, 2017 at 5:36 PM, Bogdan Gusiev <notifications@github.com>
wrote:

> Agree this is not 100% backward compatible. But if you would design the
> system this way, if you would start it from scratch, this is a good sign to
> review this as a direction to go. We can search for solutions if backward
> compatibility is the only concern.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/rails/rails/pull/29092#issuecomment-301512878>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAAKtWk-b7rCch8xhvGq6jR5Ej2Cfgqoks5r6HDvgaJpZM4NbLH7>
> .
>

kaspth commented:
> you turn over your entire cache allowance frequently enough

Just ask ol' daddy-o for some bigger smackeroos then, sonny! 😄

kaspth commented:
> Unfinished: Dealing with multi_get/fetch

Are you tackling this as part of this PR or intending that for a later one?

kaspth commented on activesupport/lib/active_support/cache.rb:
```diff
@@ -232,6 +232,11 @@ def mute
       # new value. After that all the processes will start getting the new value.
       # The key is to keep <tt>:race_condition_ttl</tt> small.
       #
+      # Setting <tt>:version</tt> will verify that the cache stored in the <tt>name</tt>
```
This took me some tries to understand, so I took a dab at slimming it:

```ruby
# Passing a <tt>:version</tt> verifies the cache stored under <tt>name</tt>
# is of the same version. nil is returned on mismatches despite contents.
# This feature is used to support recyclable cache keys.
```

dhh replied:
Like that.

kaspth commented on activesupport/lib/active_support/cache.rb:
```diff
@@ -307,17 +313,30 @@ def fetch(name, options = nil)
       # the cache with the given key, then that data is returned. Otherwise,
       # +nil+ is returned.
       #
+      # As with fetch, the data is only returned if it has not expired per the
```
Why mention `fetch`? Also confusing when put together with `Fetches data from the cache` which is how the `read` doc starts.

kaspth commented on activesupport/lib/active_support/cache.rb:
```diff
@@ -307,17 +313,30 @@ def fetch(name, options = nil)
       #
+      # As with fetch, the data is only returned if it has not expired per the
+      # <tt>:expires_in<tt> option, and, if a <tt>:version</tt> parameter is passed
+      # to <tt>read</tt>, if it matches the <tt>:version</tt> it was written with.
```
Also took me some tries. How about:

```ruby
# Any found entry is only returned if it has not expired per the
# <tt>:expires_in</tt> option and when passed a
# <tt>:version</tt> parameter if the entry's version matches that.
```

dhh replied:
Make it so.

kaspth commented on activesupport/lib/active_support/cache.rb:
```diff
@@ -307,17 +313,30 @@ def fetch(name, options = nil)
+            elsif entry.mismatched?(options[:version])
+              if payload
+                payload[:hit]      = false
+                payload[:mismatch] = "#{entry.version} != #{options[:version]}"
```
I assume we'll be logging this in the Action View log subscribing, right?

dhh replied:
Yes, I have a subscriber there to output this. It's a bit annoying that this is disconnected from the general instrumentation of read/write, because it means it has to be logged as a separate line. Would be much nicer if the read log line could say whether there was a version hit or not. Thoughts on how?

kaspth replied:
> Yes, I have a subscriber there to output this. It's a bit annoying that this is disconnected from the general instrumentation of read/write, because it means it has to be logged as a separate line. Would be much nicer if the read log line could say whether there was a version hit or not. Thoughts on how?

@dhh why does it have to be logged as a separate line?

I guess we could make this type of logging have tags as well, except they're tailing unlike the standard tagged logging.

Something like:

```ruby
payload.tags[:hit]             = false
payload.tags[:mismatch] = "#{entry.version} != #{options[:version}"
# Later: …cache/key… [hit: false][mismatch: 123 != 456]
```

Would require `payload` being more than a Hash though.

dhh replied:
Yeah, right now we're not using this information. Don't think we need it at the moment given the fact that all fragment keys are now using versions. I'll nix it.

Could you have a look at the cache hit/miss issue?

kaspth replied:
Yeah, I'd rather hit that issue out of the park than miss it 🤓

kaspth reviewed (approved):
Mostly just found the docs a bit hard to read, 👍

Add a changelog entry too 😉

bogdan commented:
@dhh do you plan to change `ActiveRecord::Base#cache_key` method? I think you can not move forward without it and it will be backward incompatible anyway (or at least generate the same type of problem as in my suggestion). Anyway I made the `cache_version` method in the way imagine it: https://github.com/rails/rails/pull/29107

dhh commented:
I don't intend to change that method. What the CacheHelper#vcache method I will propose shortly does is check arity on that method and call it with #cache_key(:without_version). That relies on the fact that the first parameter to AR::Base#cache_key specifies the timestamp field to be used and if there's no match, it won't include a timestamp.

dhh commented:
@kaspth Any thoughts on how we could make this work with multiget? I'm thinking that this low-level API could just be `Rails.cache.fetch("a", "b", versions: [ 1, 2 ])`. Then we'd need to do the right thing for `cache: true` in the `render collection` call.

bogdan commented:
That sounds pretty sad because `cache_key` is only used for `ActiveSupport::Cache` implementation (correct me if I am wrong). It means that `cache_key` argument and timestamp will be an artifact we would need to support.

In the meanwhile the behavior of `CacheHelper#cache` will be changed anyway. It may affect Apps in the same way the change in `CacheStore#fetch` would.


I would goal for a more idealistic solution at least long term and change the `cache_key` method to remove the timestamp.

Apps relaying on `cache_key` to include timestamp would not be necessary as `cache_key` is designed to be defined per app per model.

Here is the way one can maintain the old behavior basically forever:

``` ruby
module DeprecatedCacheKey
  def cache_key
     # old cache key implementation
  end
  def cache_version
    nil
  end
end
# To maintain the old behavior temporary
ActiveRecord::Base.send(:include, DeprecatedCacheKey)
# To migrate models one by one relaying on old behavior: 
Post.send(:include, DeprecatedCacheKey)
```

dhh commented:
I think it's quite reasonable to leave `CacheHelper#cache` in place and use a second helper, like `CacheHelper#vcache` that uses the new version-based strategy. This would mean that people can adopt as they please.

`cache_key` is used in a variety of ways, not just by ActiveSupport::Cache. I don't think blowing backwards compatibility is necessary or would buy us very much.

dhh commented:
Although, we could also consider that if cache_version is present on the model, then cache_key does not include the version. That would be backwards compatible.

bogdan commented:
> I think it's quite reasonable to leave CacheHelper#cache in place and use a second helper, like CacheHelper#vcache that uses the new version-based strategy. This would mean that people can adopt as they please.

When I see `vcache` besides `cache`, I imagine how I would deliver the knowledge  to the team of 20 people, where not all of them are Sr developers. Ideally I want `vcache` to be the only thing used. I don't imagine `vcache` and `cache` working together in apps that were generated on 5.2+. `cache versioning: true` looks like a better idea anyway.

> Although, we could also consider that if cache_version is present on the model, then cache_key does not include the version. That would be backwards compatible.

That is a good step forward. We should definitely do that in case we go for the `cache_version` method support. 
The problem of canonic `cache_version` method in each app would be a problem.
It can be simplified to the following:

``` ruby
class ApplicationRecord < AR::Base
  def cache_version
    ActiveRecord::Base.cache_version(self)
  end
end
# or maybe even
ApplicationRecord.cache_version = true
```

dhh commented:
I would bake #cache_version into ActiveModel and then perhaps provide a
top-level setting for making it return nil for backwards compatibility.

Let's see how far we can get with all this. If we can arrive at a place
where we have full backwards compatibility, full support for multiget etc,
then maybe we won't need a new vcache method.

On Tue, May 16, 2017 at 2:05 PM, Bogdan Gusiev <notifications@github.com>
wrote:

> I think it's quite reasonable to leave CacheHelper#cache in place and use
> a second helper, like CacheHelper#vcache that uses the new version-based
> strategy. This would mean that people can adopt as they please.
>
> When I see vcache besides cache, I imagine how I would deliver the
> knowledge to the team of 20 people, where not all of them are Sr
> developers. Ideally I want vcache to be the only thing used. I don't
> imagine vcache and cache working together in apps that were generated on
> 5.2+. cache versioning: true looks like a better idea anyway.
>
> Although, we could also consider that if cache_version is present on the
> model, then cache_key does not include the version. That would be backwards
> compatible.
>
> That is a good step forward. We should definitely do that in case we go
> for the cache_version method support.
> The problem of canonic cache_version method in each app would be a
> problem.
> It can be simplified to the following:
>
> class ApplicationRecord < AR::Base
>   def cache_version
>     ActiveRecord::Base.cache_version(self)
>   endend# or maybe evenApplicationRecord.cache_version = true
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rails/rails/pull/29092#issuecomment-301761016>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAAKtU4UbCzMJl3e10HwsI7o5_fuP6mbks5r6ZDwgaJpZM4NbLH7>
> .
>

Commit:
Support #cache_version and multiget when that's used

bogdan commented on activesupport/lib/active_support/cache.rb:
```diff
@@ -547,6 +579,10 @@ def normalize_key(key, options)
         end
 
+        def normalize_version(key, options)
+          options[:version] || key.try(:cache_version)
```
You need to support more advanced array  keys like:

``` ruby
fetch(["preview", post])
fetch(["preview", post, post.author])
```

There is an implementation in `expand_cache_version` here: https://github.com/rails/rails/pull/29107/files#diff-438394335b9c1ce6ec4f67a407f50a42R89

Commit:
Expand the cache version properly

Commit:
Add the cache_version expected by ActiveSupport::Cache when using recycling cache keys

dhh commented:
I think the #cache_version as the toggle has legs. Just extended it to Active Record in a backwards compatible form 👍.

christos commented on activerecord/lib/active_record/integration.rb:
```diff
@@ -13,6 +13,15 @@ module Integration
+
+      ##
+      # :singleton-method:
+      # Indicates whether to use a stable #cache_key method that is accompaigned
```
`s/accompaigned/accompanied/`

bogdan commented on activesupport/lib/active_support/cache.rb:
```diff
@@ -85,6 +85,14 @@ def expand_cache_key(key, namespace = nil)
 
+      def expand_cache_version(key)
+        case
+        when key.respond_to?(:cache_version) then key.cache_version
```
We need to have a consistent behavior for the cache version. `cache_key` can only be a string at the end (meaning that String is always put into cache store). Maybe we need the same for version. In this case we always need to call `to_param` here.

bogdan commented on activerecord/lib/active_record/integration.rb:
```diff
@@ -52,25 +61,47 @@ def to_param
       else
-        timestamp = if timestamp_names.any?
-          max_updated_column_timestamp(timestamp_names)
+        if cache_version
```
When `cache_key` is called with argument, the `cache_version` setting should be ignored.

I am not sure why `cache_key` has an argument: I have checked the code and there is no calls to `cache_key` with arguments neither from AS::Cache nor from other places. I would deprecate the argument as part of this PR.

dhh replied:
The argument is for calling like `<% cache [ person.cache_key(:bio_updated_at) ] %>`, but yes, not a common usage.

bogdan replied:
Got it. From this usecase: when `cache_key` is called with argument, the `cache_version` setting should be ignored.

bogdan commented on activesupport/lib/active_support/cache.rb:
```diff
@@ -85,6 +85,14 @@ def expand_cache_key(key, namespace = nil)
         expanded_cache_key
       end
 
+      def expand_cache_version(key)
```
This method is currently public. I am not sure why would anyone call it explicitly. I would make it private. Otherwise it needs to be documented.

Commit:
No need for a public cache version expander, but normalize it

dhh commented:
Current problem is that to use versioning via views, it goes CacheHelper#cache -> Fragments#write_fragment -> ActiveSupport::Cache.expand_cache_key -> ActiveSupport::Cache::Store#write. Well, in the expand_cache_key step we currently convert the key to a string, which means that ActiveSupport::Cache::Store#write can't introspect it for #cache_version. So need to retain the array-based key all the way from the top to the store, which I'm having some trouble trying to do in a backwards compatible manner.

Commit:
Keep the timestamp cache version in the key if its explicitly specified

Commit:
Spelling

Commit:
Separate recursive method like with expanded_key

bogdan commented:
### Variant 1

Push the `:version` option explicitly into `cache_store.write`

```
cache_store.write(key, content, options.reverse_merge(version: expand_cache_version(key))
```

### Variant 2

From the first glance: there should be 2 version of the `fragment_cache_key`: with `expand_cache_key` call and without it. The one with it, should be left to be called by the end user like it is now, the one without it should be internal and called from `write_fragment` and similar methods.

dhh commented:
Yeah, I'm pursuing variant 2 right now.

On Wed, May 17, 2017 at 12:57 PM, Bogdan Gusiev <notifications@github.com>
wrote:

> Variant 1
>
> Push the :version option explicitly into cache_store.write
>
> cache_store.write(key, content, options.reverse_merge(version: expand_cache_version(key))
>
> Variant 2
>
> From the first glance: there should be 2 version of the fragment_cache_key:
> with expand_cache_key call and without it. The one with it, should be
> left to be called by the end user like it is now, the one without it should
> be internal and called from write_fragment and similar methods.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rails/rails/pull/29092#issuecomment-302056190>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAAKtQsOHJHVhwjoS21BgrbkICRIOUlsks5r6tKogaJpZM4NbLH7>
> .
>

Commit:
Allow fragment accessors to pass original keys as an array to store rather than an expanded string

This allows the store to interrogate the individual elements of the key
array for #cache_version’s.

Commit:
Style

Commit:
Not all records have an updated_at column

Commit:
Remove debug information

Commit:
Prevent elements without versions from combining via nils

Commit:
Go with DB-like version string

Commit:
Enhance the fragment name to include the virtual path for easier debugging

Commit:
Expand key to get a string now that it comes as default in array form

Commit:
Use seconds since epoch. No spaces, just a number.

Commit:
Only include digest if there is one

Commit:
Fix for new key format

Commit:
Unused

Commit:
Drop new testing table as well

Commit:
Doc updates

dhh commented:
Okay, now there's a good baseline. It covers multiget, it's backwards compatible, and the API is neat as long as you implement #cache_version 👍

Commit:
Don't use Ruby 2.2 features yet

Commit:
Rubocop appeasement

bogdan commented:
There is one more usage of `expand_cache_key`: https://github.com/rails/rails/blob/bc4781583d9db237dd928b01743c6db7596a36b3/actionpack/lib/abstract_controller/caching.rb#L57

I tried to comment out this method locally and all tests passed. Maybe it just can be deleted.

dhh commented:
That's for caching stuff in the controller. It should be kept, but we
should add tests for it.

On Wed, May 17, 2017 at 4:04 PM, Bogdan Gusiev <notifications@github.com>
wrote:

> There is one more usage of expand_cache_key: https://github.com/rails/
> rails/blob/bc4781583d9db237dd928b01743c6db7596a36b3/actionpack/lib/
> abstract_controller/caching.rb#L57
>
> I tried to comment out this method locally and all tests passed. Maybe it
> just can be deleted.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rails/rails/pull/29092#issuecomment-302100037>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAAKtSwvh5QdGZhA2s8Fod7f9HcrVGzpks5r6v5sgaJpZM4NbLH7>
> .
>

Commit:
Missing word

bogdan commented:
Should it be updated to remove the direct `expand_cache_key` call?

bogdan commented on actionpack/lib/action_controller/log_subscriber.rb:
```diff
@@ -60,9 +60,9 @@ def unpermitted_parameters(event)
         def #{method}(event)
           return unless logger.info? && ActionController::Base.enable_fragment_cache_logging
-          key_or_path = event.payload[:key] || event.payload[:path]
+          key         = ActiveSupport::Cache.expand_cache_key(event.payload[:key] || event.payload[:path])
```
`RAILS_CACHE_ID` var is weird: currently it only applies to caches made through actionpack, but not directly at `Rails.cache`.

The current problem with that can be here because you add `RAILS_CACHE_ID` to `payload[:key]` explicitly now and then expand it with same ENV var prefix again.


```
$ RAILS_CACHE_ID=22 rails c
Loading development environment (Rails 5.0.2)
development >> ActiveSupport::Cache.expand_cache_key('qq')
=> "22/qq/1"
development >> Rails.cache.fetch("qq") { 1}
=> 1
development >> Rails.cache.read(ActiveSupport::Cache.expand_cache_key('qq'))
=> nil
```
I think it is a bug and all direct calls to `Rails.cache` should also incorporate `RAILS_CACHE_ID`.

dhh replied:
Agree. I'd rather make it a direct setting for Rails.cache, but that's for a later PR.

dhh commented:
Yeah, that sounds better to lift it up higher. Do take a stab at that.

On Wed, May 17, 2017 at 4:21 PM, Bogdan Gusiev <notifications@github.com>
wrote:

> *@bogdan* commented on this pull request.
> ------------------------------
>
> In actionpack/lib/action_controller/log_subscriber.rb
> <https://github.com/rails/rails/pull/29092#discussion_r117014424>:
>
> > @@ -60,9 +60,9 @@ def unpermitted_parameters(event)
>        class_eval <<-METHOD, __FILE__, __LINE__ + 1
>          def #{method}(event)
>            return unless logger.info? && ActionController::Base.enable_fragment_cache_logging
> -          key_or_path = event.payload[:key] || event.payload[:path]
> +          key         = ActiveSupport::Cache.expand_cache_key(event.payload[:key] || event.payload[:path])
>
> RAILS_CACHE_ID var is weird: currently it only applies to caches made
> through actionpack, but not directly at Rails.cache.
>
> The current problem with that can be here because you add RAILS_CACHE_ID
> to payload[:key] explicitly now and then expand it with same ENV var
> prefix again.
>
> $ RAILS_CACHE_ID=22 rails c
> Loading development environment (Rails 5.0.2)
> development >> ActiveSupport::Cache.expand_cache_key('qq')
> => "22/qq/1"
> development >> Rails.cache.fetch("qq") { 1}
> => 1
> development >> Rails.cache.read(ActiveSupport::Cache.expand_cache_key('qq'))
> => nil
>
> I think it is a bug and all direct calls to Rails.cache should also
> incorporate RAILS_CACHE_ID.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rails/rails/pull/29092#pullrequestreview-38682062>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAKtcObNf4LlKDrAjBOPf_Ov4NIty2Lks5r6wKCgaJpZM4NbLH7>
> .
>

dhh commented:
Problem is that some of the logic for doing it right with versions and
incorporating all the RAILS_CACHE_ID etc lives further down. Think it's a
bigger extraction. Shouldn't hold this up. We can come back to it. Don't
think it's very used.

On Wed, May 17, 2017 at 4:13 PM, Bogdan Gusiev <notifications@github.com>
wrote:

> Should it be updated to remove the direct expand_cache_key call?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rails/rails/pull/29092#issuecomment-302102616>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAAKtXouecZItYQAURyEt_GlhyQRQySyks5r6wBvgaJpZM4NbLH7>
> .
>

Commit:
Move documentation

rafaelfranca commented on actionpack/test/controller/caching_test.rb:
```diff
@@ -53,12 +51,17 @@ def setup
   end
 
   def test_fragment_cache_key
-    assert_equal "views/what a key", @controller.fragment_cache_key("what a key")
```
Maybe keep the tests for `fragment_cache_key` and add new ones for `combined_fragment_cache_key`? is deprecated we need to keep its tests otherwise the behavior can change and we don't notice.

dhh replied:
Yeah, good idea.

rafaelfranca commented on actionpack/test/controller/caching_test.rb:
```diff
@@ -53,12 +51,17 @@ def setup
-                  @controller.fragment_cache_key(controller: "fragment_caching_test", action: "some_action")
+    assert_equal [ :views, "what a key" ], @controller.combined_fragment_cache_key("what a key")
+    assert_equal [ :views, "test.host/fragment_caching_test/some_action" ],
+      @controller.combined_fragment_cache_key(controller: "fragment_caching_test", action: "some_action")
```
Need tests for `RAILS_CACHE_ID` and `RAILS_APP_VERSION`

rafaelfranca commented on activerecord/lib/active_record/integration.rb:
```diff
@@ -13,6 +13,15 @@ module Integration
+      #
+      # This is +false+, by default until Rails 6.0.
+      class_attribute :cache_versioning, instance_writer: false
+      self.cache_versioning = false
```
Do we want new 5.2 applications to use this as true by default?

dhh replied:
Yes, @matthewd has a checklist of what's needed to switch over.

rafaelfranca commented on activerecord/test/cases/integration_test.rb:
```diff
@@ -177,4 +177,50 @@ def test_cache_key_when_named_timestamp_is_nil
+    developer     = Developer.first
+    first_version = developer.cache_version
+
+    travel 10.seconds
```
We need to use the block version or `travel_back` otherwise this test will leak.

dhh replied:
I thought there was a teardown hook to reset? Anyway, can do a block.

rafaelfranca replied:
We don't have a teardown hook right now. But we can add it.

dhh replied:
That'd be great.

rafaelfranca commented on activesupport/lib/active_support/cache.rb:
```diff
@@ -517,6 +539,17 @@ def merged_options(call_options)
           end
         end
 
+
```
Extra blank line here.

dhh replied:
Intentional, separating related pairs of methods.

rafaelfranca replied:
hmmm. Rubocop will remove it when we run the autocorrect though

dhh replied:
Let's nix that Rubocop rule then. I often use an extra return to separate logical units of related methods.

rafaelfranca commented on activesupport/lib/active_support/cache.rb:
```diff
@@ -555,6 +590,7 @@ def instrument(operation, key, options = nil)
           ActiveSupport::Notifications.instrument("cache_#{operation}.active_support", payload) { yield(payload) }
         end
 
+
```
Extra blank line here.

dhh replied:
Intentional, separating related pairs of methods.

Commit:
Fix tests for new fragment key format

Commit:
Test env fragment cache keys

Commit:
Don't pollute cross tests

Commit:
Fix new Rails 5.1 default for form_with_generates_remote_forms

CC @matthewd @kaspeth

Commit:
Set new default for cache versioning via 5.2 defaults

Commit:
Close the if

Commit:
Start the new setting commented out

Commit:
Fix test

Commit:
5.1 default is to generate remote forms

Commit:
Fix tests for new cache key format

Commit:
Fix tests

Commit:
Appease Rubocop

Commit:
Fix the remaining bits for switching to new framework defaults

Commit:
Update documentation to match new caching style

dhh commented:
This cache_version might well not be used just for the cache
implementation. People may well want to do their own stuff with it, and as
such the versions should be easily comparable, like person.cache_version >
older_person.cache_version.

On Thu, May 18, 2017 at 4:58 PM, Bogdan Gusiev <notifications@github.com>
wrote:

> *@bogdan* commented on this pull request.
> ------------------------------
>
> In activerecord/lib/active_record/integration.rb
> <https://github.com/rails/rails/pull/29092#discussion_r117195651>:
>
> >          end
>        end
>      end
>
> +    # Returns a cache version that can be used together with the cache key to form
> +    # a recyclable caching scheme. By default, the #updated_at column is used for the
> +    # cache_version, but this method can be overwritten to return something else.
> +    #
> +    # Note, this method will return nil if ActiveRecord::Base.cache_versioning is set to
> +    # +false+ (which it is by default until Rails 6.0).
> +    def cache_version
> +      try(:updated_at).try(:to_i) if cache_versioning
>
> As we convert all versions to string at the low level, it makes sense to
> call to_s here too so that direct call to cache_version would return the
> actual version value that will be put to the cache store.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rails/rails/pull/29092#pullrequestreview-38879739>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAKtTwej8cQX4FFCHJspNgcb4ycj8Q7ks5r7FyzgaJpZM4NbLH7>
> .
>

dhh commented:
CI is complaining of this failure:

<img width="589" alt="screen shot 2017-05-18 at 5 15 27 pm" src="https://cloud.githubusercontent.com/assets/2741/26209624/9bb5606e-3bed-11e7-9710-79f8dcf8f982.png">

I can't replicate that:

<img width="1031" alt="screen shot 2017-05-18 at 5 17 27 pm" src="https://cloud.githubusercontent.com/assets/2741/26209723/e30f695a-3bed-11e7-9f57-b44361030a19.png">

So not going to let that hold up the merge.

Commit:
Changelog entries

Commit:
Merge branch 'master' into versioned-cache-entries

mikeycgto commented on railties/lib/rails/generators/rails/app/app_generator.rb:
```diff
@@ -121,7 +121,7 @@ def config_when_updating
       rack_cors_config_exist = File.exist?("config/initializers/cors.rb")
       assets_config_exist = File.exist?("config/initializers/assets.rb")
-      new_framework_defaults_5_1_exist = File.exist?("config/initializers/new_framework_defaults_5_1.rb")
+      new_framework_defaults_5_2_exist = File.exist?("config/initializers/new_framework_defaults_5_2.rb")
```
`new_framework_defaults_5_2_exist` is now an unused variable. It is getting reported as a warning from the test:

```
rails/railties/lib/rails/generators/rails/app/app_generator.rb:124: warning: assigned but unused variable - new_framewo
rk_defaults_5_2_exist
```

dhh replied:
Fixed in e4ce81c234 👍

Dorian commented:
For people hoping on the discussion like, here is how the cache keys are generated: https://stackoverflow.com/a/23347985/407213

tomrossi7 commented:
I'm having trouble migrating to edge Rails and noticed my cache keys no longer include timestamps. Will the discussion on this thread require updates to the [documentation on caching here](http://edgeguides.rubyonrails.org/caching_with_rails.html#russian-doll-caching)?

gingerlime commented:
I noticed some unexpected behaviour related to this change, when used with Active Record relations (e.g. `cache Product.all`). 

The cache_key implementation works great, but when [`normalize_version`](https://github.com/rails/rails/blob/09ca939eabc8b0cda865ecf2893c48800a7c51e3/activesupport/lib/active_support/cache.rb#L668-L678) is executed, the relation does not respond to `cache_version`, and therefore ends up being expanded with [`to_a`](https://github.com/rails/rails/blob/09ca939eabc8b0cda865ecf2893c48800a7c51e3/activesupport/lib/active_support/cache.rb#L676). So essentially, calling `Product.all.to_a` and then for each object calling `cache_version`, which returns `nil`. This obviously has significant performance repercussions, which defeats the benefits of caching in lots of cases.

I can work around it by using something like `cache ActiveSupport::Cache.expand_cache_key(Product.all)`, but this feels a bit dirty. See [SO question](https://stackoverflow.com/q/53222582/305019)

This is, by the way, without versioning switched on.

It's probably an anti-pattern to use AR relations as cache keys in views(?), but this still caught me by surprise. It works and produces the right cache key, but with a pretty hefty and unexpected side-effect.

Is there a plan to implement `cache_version` on Active Record relations? Or what's the recommended way to deal with this? (sorry if I'm posting in the wrong place, but it's directly connected to this change as far as I can tell).
